core/src/main/java/io/grpc/Metadata.java

```diff
@@ -449,7 +449,8 @@ private static String validateName(String n) {
 
     private Key(String name) {
       this.originalName = checkNotNull(name);
-      this.name = validateName(this.originalName.toLowerCase(Locale.ROOT));
+      // Intern the result for faster string identity checking.
+      this.name = validateName(this.originalName.toLowerCase(Locale.ROOT)).intern();
```

Comment:
Don't you also have to update `Key.equals` for that to work? Also wouldn't it be sufficient to do the interning in `Key.equals` when comparing?